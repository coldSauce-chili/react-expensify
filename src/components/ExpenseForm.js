import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'


const now = moment()
console.log(now.format('MMM Do, YYYY'))


export default class ExpenseForm extends React.Component {

    // Mead claimed that Props cannot be access without using constructor
    // from video: Section 11 - Wiring Up Edit Expense
    // Debunk below
    constructor(props) {
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ""
        }
    }

    // THIS IS A VALID SYNTAX!!
    //  you can access props by using this.props
    // no need constructor()
    //
    // state = {
    //     description: this.props.expense ? this.props.expense.description : '',
    //     note: this.props.expense ? this.props.expense.note : '',
    //     amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    //     createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    //     calendarFocused: false,
    //     error: ""
    // }




    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
        // console.log(this.state.description)
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
            // console.log(this.state.amount)

        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "input description and amount" }))
        } else {
            this.setState(() => ({ error: "" }))
            console.log('submitted')
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <form className='form' onSubmit={this.onSubmit}>
                {this.state.error && <p className='form__error'>{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="description"
                    autoFocus
                    className='text-input'
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="amount"
                    className='text-input'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => (false)}
                />
                <textarea
                    placeholder="notes"
                    className='textarea'
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className='button'>Add expense</button>
                </div>

            </form>
        )

    }
} 