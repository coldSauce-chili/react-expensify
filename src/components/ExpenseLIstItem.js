//ecxport stateless coponent
// redner description, amout date crated

import React from 'react'
import { Link } from 'react-router-dom'

const ExpenseListItems = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`edit/${id}`}>
            <h4>description: {description}</h4>
        </Link>
        <h5>amount: {amount}</h5>
        <h6>date: {createdAt}</h6>

    </div>
)

export default ExpenseListItems
