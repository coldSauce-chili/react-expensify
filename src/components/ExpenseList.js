import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseLIstItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>no expenses</p>
            ) : (
                    props.expenses.map((expense) => (
                        <ExpenseListItem
                            key={expense.id}
                            {...expense}
                        />
                    ))
                )
        }

    </div>
)


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)




//the 'below' is how the 'above' works

// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses
//     }
// })(ExpenseList)

// export default ConnectedExpenseList