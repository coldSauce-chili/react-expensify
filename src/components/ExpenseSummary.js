import React from 'react'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'
import totalExpense from '../selectors/expenses-total'
import numeral from 'numeral'


export const ExpenseSummary = (props) => {

    const expensesCount = props.expenses.length
    const total = numeral(totalExpense(props.expenses) / 100).format('$0,0.00')

    return (
        <div>
            <h4>viewing {expensesCount} expenses totalling {total} </h4>
        </div>
    )

}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)