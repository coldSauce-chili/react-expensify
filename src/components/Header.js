import React from 'react'
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom'

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink activeClassName="is-active" to="/" exact={true}>dash</NavLink>
        <NavLink activeClassName="is-active" to="/create">create</NavLink>
        <NavLink activeClassName="is-active" to="/help">help</NavLink>
    </header>
)

export default Header