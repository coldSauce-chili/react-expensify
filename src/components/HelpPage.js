import React from 'react'

const HelpPage = () => (
    <div>
        this is from my heeeeeeeelp expense compoent
    </div>
)

export default HelpPage