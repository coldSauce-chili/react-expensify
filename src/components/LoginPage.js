import React from 'react'
import { connect } from 'react-redux'
import { startLogin } from '../actions/auth'

// class LoginPage extends React.Component {
//     render() {
//         return (
//             <div>
//                 <button>my button</button>
//             </div>
//         )
//     }
// }

export const LoginPage = ({ startLogin }) => (
    <div>
        <button onClick={startLogin}>my button</button>
    </div>
)


const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
})
export default connect(undefined, mapDispatchToProps)(LoginPage)