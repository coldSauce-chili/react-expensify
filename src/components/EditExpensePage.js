import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {

    onRemoveExpense = (expense) => {
        this.props.onRemoveExpense(expense)
        this.props.history.push('/')
    }

    onEditExpense = (id, expense) => {
        this.props.onEditExpense(id, expense)
        this.props.history.push('/')
    }


    render() {
        return (
            <div>
                editing the expens with the ID of {this.props.match.params.id}
                <ExpenseForm
                    expense={this.props.expense}

                    onSubmit={(expense) => {
                        this.onEditExpense(this.props.expense.id, expense)
                    }}
                />
                <button onClick={() => { this.onRemoveExpense(this.props.expense) }}>
                    remove
            </button>

            </div>
        )
    }
}
// const EditExpensePage = (props) => {
//     console.log(props)
//     return (
//         <div>
//             editing the expens with the ID of {props.match.params.id}
//             <ExpenseForm
//                 expense={props.expense}

//                 onSubmit={(expense) => {
//                     props.dispatch(editExpense(props.expense.id, expense))
//                     props.history.push('/')
//                 }}
//             />
//             <button onClick={remove}>remove
//             </button>

//         </div>
//     )
// }


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => {

    return {
        onRemoveExpense: (expense) => {
            dispatch(startRemoveExpense(expense))
            // history.push('/')
        },
        onEditExpense: (id, expense) => {
            dispatch(editExpense(id, expense))
            // history.push('/')
            console.log('this is the update data')
            console.log()
        }
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage) 