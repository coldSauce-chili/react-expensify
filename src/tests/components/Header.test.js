import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { Header } from '../../components/Header'

test('should render Head correctly', () => {
    const wrapper = shallow(<Header startLogout={() => { }} />)

    expect(wrapper).toMatchSnapshot()

})

test('should call startLogOut on button click', () => {

    const startLogOutSpy = jest.fn()
    const wrapper = shallow(<Header startLogout={startLogOutSpy} />)

    wrapper.find('button').simulate('click')

    expect(startLogOutSpy).toHaveBeenCalled()

})



    // expect(wrapper.find('h1').length).toBe(1)

    // const renderer = new ReactShallowRenderer()
    // renderer.render(<Header />)

    // expect(renderer.getRenderOutput()).toMatchSnapshot()
    // console.log(renderer.getRenderOutput())