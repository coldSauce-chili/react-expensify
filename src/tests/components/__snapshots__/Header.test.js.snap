// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Head correctly 1`] = `
<header>
  <h1>
    Expensify
  </h1>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    exact={true}
    to="/"
  >
    dash
  </NavLink>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    to="/create"
  >
    create
  </NavLink>
  <NavLink
    activeClassName="is-active"
    ariaCurrent="true"
    to="/help"
  >
    help
  </NavLink>
</header>
`;
